tsl/src/continuous_aggs/planner.c

```diff
@@ -0,0 +1,425 @@
+	if (node == NULL)
+		return false;
+
+	if (IsA(node, FuncExpr))
+	{
+		FuncExpr *funcExpr = castNode(FuncExpr, node);
+
+		/* Handle watermark function */
+		if (context->watermark_function_oid == funcExpr->funcid)
+		{
+			/* No coalesce expression found so far, we are not interested in this expression */
+			if (context->parent_coalesce_expr == NULL)
+			{
+				context->valid_query = false;
+				return false;
+			}
+
+			/* Our function takes exactly one argument */
+			Assert(list_length(funcExpr->args) == 1);
+
+			context->watermark_functions = lappend(context->watermark_functions, funcExpr);
+
+			/* Only on time based hypertables, we have a to_timestamp function */
+			if (context->parent_to_timestamp_func != NULL)
+			{
+				/* to_timestamp functions take only one parameter. This should be a reference to our
+				 * function */
+				Assert(linitial(context->parent_to_timestamp_func->args) == node);
+
+				context->watermark_parent_functions =
+					lappend(context->watermark_parent_functions, context->parent_to_timestamp_func);
+			}
+			else
+			{
+				/* For non int64 partitioned tables, the watermark function is wrapped into a cast
+				 * for example: COALESCE((_timescaledb_functions.cagg_watermark(11))::integer,
+				 * '-2147483648'::integer))
+				 */
+				Node *coalesce_arg = linitial(context->parent_coalesce_expr->args);
+				if (coalesce_arg != node)
+				{
+					/* Check if the watermark function is wrapped into a cast function */
+					if (!IsA(coalesce_arg, FuncExpr) || ((FuncExpr *) coalesce_arg)->args == NIL ||
+						linitial(((FuncExpr *) coalesce_arg)->args) != node)
+					{
+						context->valid_query = false;
+						return false;
+					}
+
+					context->watermark_parent_functions =
+						lappend(context->watermark_parent_functions, coalesce_arg);
+				}
+				else
+				{
+					context->watermark_parent_functions =
+						lappend(context->watermark_parent_functions, context->parent_coalesce_expr);
+				}
+			}
+		}
+
+		/* Capture the timestamp conversion function */
+		if (list_member_oid(context->to_timestamp_func_oids, funcExpr->funcid))
+		{
+			FuncExpr *old_func_expr = context->parent_to_timestamp_func;
+			context->parent_to_timestamp_func = funcExpr;
+			bool result = expression_tree_walker(node, constify_cagg_watermark_walker, context);
+			context->parent_to_timestamp_func = old_func_expr;
+
+			return result;
+		}
+	}
+	else if (IsA(node, Query))
+	{
+		/* Recurse into subselects */
+		Query *query = castNode(Query, node);
+		return query_tree_walker(query,
+								 constify_cagg_watermark_walker,
+								 context,
+								 QTW_EXAMINE_RTES_BEFORE);
+	}
+	else if (IsA(node, CoalesceExpr))
+	{
+		/* Capture the CoalesceExpr */
+		CoalesceExpr *parent_coalesce_expr = context->parent_coalesce_expr;
+		context->parent_coalesce_expr = castNode(CoalesceExpr, node);
+		bool result = expression_tree_walker(node, constify_cagg_watermark_walker, context);
+		context->parent_coalesce_expr = parent_coalesce_expr;
```

Comment:
Does this work well when you have something pathological like `coalesce(coalesce(cagg_watermark(coalesce(....`? Would be good to have some tests for such cases. It might be easier to write this a a call chain of separate functions like "find the comparison", next "find the coalesce", next "find the cagg_watermark" and so on. This way you'll limit the state space. Now you have the same function running for every level of the expression tree, it's more generic and can get into weird sequences.

Reply:
The current implementation works with your example. Note that we do some extra checks before in `could_be_realtime_cagg_query` to restrict it for a query like the realtime cagg definition that is composed by a UNION query.

Reply:
Good points. 

* Regarding the outer `coalesce(coalesce(xxxx))`: It does work since we capture the most inner `coalesce` for the replacement. I added a test case for this.

* Regarding the inner `coalesce`: We now stop walking through the query as soon as we find a non constant value of the watermark function. I also added a test case for this case.

* I decided to use the walker to find the functions instead of what you propose since you have to write down the expression you are interested in otherwise. Since we have optional cast functions and optional time conversion functions, this becomes really complex and I switched to the current approach when I prototyped the implementation. However, the proposed alternative also works, and I think it is a matter of taste how to write this down.